pilot/pkg/model/telemetry_ecds.go

```diff
@@ -0,0 +1,200 @@
+}
+
+func StatsECDSResourceName(cfg StatsConfig) string {
+	return fmt.Sprintf("istio.io/telemetry/stats/%s", cfg.String())
+}
+
+func buildHTTPTypedExtensionConfig(class networking.ListenerClass, metricsCfg []telemetryFilterConfig) []*core.TypedExtensionConfig {
+	res := make([]*core.TypedExtensionConfig, 0, telemetryFilterHandled)
+	for _, cfg := range metricsCfg {
+		switch cfg.Provider.GetProvider().(type) {
+		case *meshconfig.MeshConfig_ExtensionProvider_Prometheus:
+			if cfg.NodeType == Waypoint {
+				res = append(res, &core.TypedExtensionConfig{
+					Name: StatsECDSResourceName(StatsConfig{
+						Provider:         StatsProviderPrometheus,
+						NodeType:         cfg.NodeType,
+						ListenerClass:    class,
+						ListenerProtocol: networking.ListenerProtocolHTTP,
+					}),
+					TypedConfig: waypointStatsConfig,
+				})
+			} else {
+				if statsCfg := generateStatsConfig(class, cfg); statsCfg != nil {
+					res = append(res, &core.TypedExtensionConfig{
+						Name: StatsECDSResourceName(StatsConfig{
+							Provider:         StatsProviderPrometheus,
+							NodeType:         cfg.NodeType,
+							ListenerClass:    class,
+							ListenerProtocol: networking.ListenerProtocolHTTP,
+						}),
+						TypedConfig: statsCfg,
+					})
+				}
+			}
+		case *meshconfig.MeshConfig_ExtensionProvider_Stackdriver:
+			sdCfg := generateSDConfig(class, cfg)
+			vmConfig := ConstructVMConfig("envoy.wasm.null.stackdriver")
+			vmConfig.VmConfig.VmId = stackdriverVMID(class)
+
+			wasmConfig := &httpwasm.Wasm{
+				Config: &wasm.PluginConfig{
+					RootId:        vmConfig.VmConfig.VmId,
+					Vm:            vmConfig,
+					Configuration: sdCfg,
+				},
+			}
+
+			res = append(res, &core.TypedExtensionConfig{
+				Name: StatsECDSResourceName(StatsConfig{
+					Provider:         StatsProviderStackdriver,
+					NodeType:         cfg.NodeType,
+					ListenerClass:    class,
+					ListenerProtocol: networking.ListenerProtocolHTTP,
+				}),
+				TypedConfig: protoconv.MessageToAny(wasmConfig),
+			})
+		default:
+			// Only prometheus and SD supported currently
+			continue
+		}
+	}
+	return res
+}
+
+func buildTCPTypedExtensionConfig(class networking.ListenerClass, metricsCfg []telemetryFilterConfig) []*core.TypedExtensionConfig {
+	res := make([]*core.TypedExtensionConfig, 0, telemetryFilterHandled)
+	for _, cfg := range metricsCfg {
+		switch cfg.Provider.GetProvider().(type) {
+		case *meshconfig.MeshConfig_ExtensionProvider_Prometheus:
+			if cfg.NodeType == Waypoint {
+				res = append(res, &core.TypedExtensionConfig{
+					Name: StatsECDSResourceName(StatsConfig{
+						Provider:         StatsProviderPrometheus,
+						NodeType:         cfg.NodeType,
+						ListenerClass:    class,
+						ListenerProtocol: networking.ListenerProtocolTCP,
+					}),
+					TypedConfig: waypointStatsConfig,
+				})
+			} else {
+				if statsCfg := generateStatsConfig(class, cfg); statsCfg != nil {
```

Comment:
nit: It might be better to send something in the else clause rather than stalling the proxy.